src/Microsoft.ComponentDetection.Detectors/rust/RustCliDetector.cs

```diff
@@ -170,4 +242,195 @@ private static (string Name, string Version) ParseNameAndVersion(string nameAndV
+        CargoPackage parentPackage,
+        CargoComponent parentComponent,
+        string dependency)
+    {
+        try
+        {
+            // Extract the information from the dependency (name with optional version and source)
+            if (!ParseDependency(dependency, out var childName, out var childVersion, out var childSource))
+            {
+                // Could not parse the dependency string
+                throw new FormatException($"Failed to parse dependency '{dependency}'");
+            }
+
+            if (!packagesByName.TryGetValue(childName, out var candidatePackages))
+            {
+                throw new FormatException($"Could not find any package named '{childName}' for depenency string '{dependency}'");
+            }
+
+            // Search through the list of candidates to find a match (note that version and source are optional).
+            CargoPackage childPackage = null;
+            CargoComponent childComponent = null;
+            foreach (var (candidatePackage, candidateComponent) in candidatePackages)
+            {
+                if (childVersion != null && candidatePackage.Version != childVersion)
+                {
+                    // This does not have the requested version
+                    continue;
+                }
+
+                if (childSource != null && candidatePackage.Source != childSource)
+                {
+                    // This does not have the requested source
+                    continue;
+                }
+
+                if (childPackage != null)
+                {
+                    throw new FormatException($"Found multiple matching packages for dependency string '{dependency}'");
+                }
+
+                // We have found the requested package.
+                childPackage = candidatePackage;
+                childComponent = candidateComponent;
+            }
+
+            if (childPackage == null)
+            {
+                throw new FormatException($"Could not find matching package for dependency string '{dependency}'");
+            }
+
+            if (IsLocalPackage(childPackage))
+            {
+                // This is a dependency on a package without a source
+                return;
+            }
+
+            var detectedComponent = new DetectedComponent(childComponent);
+            seenAsDependency.Add(childPackage);
+
+            if (IsLocalPackage(parentPackage))
+            {
+                // We are adding a root edge (from a local package)
+                singleFileComponentRecorder.RegisterUsage(detectedComponent, isExplicitReferencedDependency: true);
+            }
+            else
+            {
+                // we are adding an edge within the graph
+                singleFileComponentRecorder.RegisterUsage(detectedComponent, isExplicitReferencedDependency: false, parentComponentId: parentComponent.Id);
+            }
+        }
+        catch (Exception e)
```

Comment:
We should avoid catching such generic exceptions. If we're only throwing FormatExceptions catch that

Reply:
In the majority of cases yes, but there are a few other exceptions that could mess with parsing, like out of memory exceptions or other weird edge cases, we log whatever the error is but take the safe route of catching all to not fail out of the entire detector. It's horrible practice but it matches how we do dependency parsing for our other detectors :(